### DISCUSSION TITLE
dataset weighting

### DISCUSSION THREAD

**Discussion by vince62s** [coqui-ai/TTS#1185] [Ideas]
Hello,

For finetuning, it seems relevant to overweight a smaller dataset WRT the main data used to train the model.
My suggestion would be to add a weight field in the dataset config list and to take it into account in the load_tts_samples() function.
it's simple and should make the job.
@erogol  Any insights?

**Comment by erogol** [comment]
Yeah makes sense. However, I am not sure if it would make a huge difference in performance. Any experience with that?

**Comment by vince62s** [comment]
yes it does. for now as a workaround just adding N times the same small finetuning dataset and the more I add the quicker and the higher I am reaching a high SECS.

**Reply by erogol** [reply]
Speaker sampler has already implemented in the VITS model. Would that help in this case ? @Edresson

**Comment by vince62s** [comment]
would require some changes because rigth now, language sampler and speaker sampler are mutually exclusive and here https://github.com/coqui-ai/TTS/blob/c63bb481e95bd4a1ff978947d8e3e6c0bfb4177f/TTS/tts/models/base_tts.py#L357-L363 language sampler prevails (for a good reason).

**Reply by Edresson** [reply]
@vince62s In the paper, we used only the language sampler and created a new language for the speaker that we adapted (ensuring 1/4 of the batch was composed by the speaker adapted). To do so, just edit the nn.embedding layer of language, inserting in the last position a copy of the embedding of the language you want to adapt. As the language name in config.json you need to use some name that will be in last considering the alphabetical order of the training languages (considering the training languages "en", "pt" and "fr" a good name is "spk_adpt_en" for example). We intend to enable the language and speaker balancer to work together in the near future.

**Reply by vince62s** [reply]
I know what you did Julian made that comment in another discussion. However my post here is more to say that a data balancer would be welcome to generalize weighting in both dimensions (speaker/language). Cheers.

**Reply by Edresson** [reply]
Ok, in my opinion, a dataset balancer is not the better option. It can solve the current problem (considering only speaker-balanced datasets, with the number of samples per speaker near to equal). However, in the future, we can have emotion/style embedding as well so we may need to balance by language, speaker, and emotion. So in this case make sense to just normalize between 0 and 1 the weights from language and speaker balancer and added an independent alpha for each one. After just sum the weights multiplied by the respective alphas of language and speaker balancer. In this way, we can control the influence of speaker and language in the batch balancer and easily add a new balancer if necessary in the future.  Is it make sense for you?

**Reply by vince62s** [reply]
interesting. When you mean "emotion" would that be applicable to "context style" (a bit like lovo.ai does to make a difference between a marketing prez or documentary or whatever). What I am saying, it's not always easy to make a choice between a "speaker" and s "style".
Anyway as long as there is a way to adjust data weights, it's fine.

**Reply by yiwei0730** [reply]
@Edresson  In the paper, we used only the language sampler and created a new language for the speaker that we adapted (ensuring 1/4 of the batch was composed by the speaker adapted)
I read the YourTTS paper and I see this in the adaptation training a one minutes new data, can i ask for how can i use the code in this TTS project to adapt my new dataset with the adapt training, since I didn't see how the code to chose the data 1/4 in the batch.

**Reply by vince62s** [reply]
since they add initially 3 languages and created a fourth one with the "adapted speaker" data it balanced to 1/4 automatically, but this is not a good way to do things in the long run.

**Reply by yiwei0730** [reply]
Thanks for your reply @vince62s. 
Yes, I think this is a good issue for how to use the less resource data to adapt the data. I have seen the AdaSpeech a long time, they use a conditional layer norm to create the condition for the adaption with 20 unseen speaker data. So i am so interested for how to adapting in the low resources with unseen data. 
But in the TTS code, I haven't seen the adapting training code in there, so want to ask for the author @Edresson,
By the base_tts.py code, if i want to use different speaker manager in here by multispeaker manager, so I need to use more dataset item list in the training config?

**Reply by yiwei0730** [reply]
@vince62s If it is not a good way to do thing in the long run, which methods you prefer to use, like the adaspeech using, using a novel layer or references the other papers?

**Reply by Edresson** [reply]
Hi @vince62s  In YourTTS paper, we used only the language sampler and created a new language for the speaker that we adapted (ensuring 1/4 of the batch was composed by the speaker adapted). To do so, just edit the nn.embedding layer of language, inserting in the last position a copy of the embedding of the language you want to adapt. As the language name in config.json you need to use some name that will be in last considering the alphabetical order of the training languages (considering the training languages "en", "pt" and "fr" a good name is "spk_adpt_en" for example).This approach is far to be optima, for this reason we implemented a way to balancer the language and speaker in a controllable way  in this PR #1216 

I didn't understand this part "By the base_tts.py code,  if i want to use different speaker manager in here by multispeaker manager, so I need to use more dataset item list in the training config?" Could you please elaborate?
